src/type-checker.ts

```diff
@@ -42,196 +38,232 @@ function visitSerial (nodes: AstNode[], env: TypeEnv, nonGeneric: Set<Type>): Ty
   , initialState)
 }
 
-function visitAll (nodes, rootEnv, nonGeneric) {
-  return nodes.reduce(([nodes, env], node) => {
+// Visit a list of nodes, each in the returned env of the previous step.
+// Return an array of their resolved types and the last environment.
+// TODO DRY visitSerial to use this function
+// TODO static type for `nodes` - the callback throws type errors for nodes: AstNode[]
+function visitAll (nodes: any[], rootEnv: TypeEnv, nonGeneric: Set<any>): [TypedNode[], TypeEnv] {
+  const initialState: [TypedNode[], TypeEnv] = [[], rootEnv]
+
+  return nodes.reduce(([nodes, env]: [TypedNode[], TypeEnv], node) => {
     const [outNode, outEnv] = visit(node, env, nonGeneric)
     return [[...nodes, outNode], outEnv]
-  }, [[], rootEnv])
+  }, initialState)
 }
 
-function visit (node: AstNode, env: TypeEnv, nonGeneric: Set<Type>) {
-  const visitor = visitors[node.type]
-  return visitor(node, env, nonGeneric)
+function visit (node: AstNode, env: TypeEnv, nonGeneric: Set<Type>): TypeCheckResult<TypedNode> {
+  // console.log(`TRACE typecheck: ${node.type}\n${JSON.stringify(node)}`)
+
+  switch (node.type) {
+    case 'Program':
+      return visitProgram(node, env, nonGeneric)
+    case 'Def':
+      return visitDef(node, env, nonGeneric)
+    case 'Name':
+      return visitName(node, env, nonGeneric)
+    case 'Numeral':
+      return visitNumeral(node, env)
+    case 'Bool':
+      return visitBool(node, env)
+    case 'Str':
+      return visitStr(node, env)
+    case 'Call':
+      return visitCall(node, env, nonGeneric)
+    case 'Array':
+      return visitArray(node, env, nonGeneric)
+    case 'DestructuredArray':
+      return visitDestructuredArray(node, env, nonGeneric)
+    case 'Fn':
+      return visitFn(node, env, nonGeneric)
+    case 'If':
+      return visitIf(node, env, nonGeneric)
+    default:
+      throw new Error(`Uncrecognised AST node type: ${node.type}`)
+  }
 }
 
-const visitors: { [nodeType: string]: TypeCheckVisitor<TypedNode> } = {
-  Program (node: AstProgramNode, env, nonGeneric) {
-    const [expressions, finalEnv] = visitAll(node.expressions, env, nonGeneric)
-    const programType = typeOf(last(expressions))
+function visitProgram (node: AstProgramNode, env, nonGeneric): TypeCheckResult<TypedProgramNode> {
+  const [expressions, finalEnv] = visitAll(node.expressions, env, nonGeneric)
+  const programType = typeOf(last(expressions))
 
-    const typedNode = typed({ ...node, expressions }, programType)
-    return [typedNode, finalEnv]
-  },
+  const typedNode = { ...node, expressions, exprType: programType }
```

Comment:
Inlining `typed` so that TypeScript can use `exprType` as a tag for the `TypedNode` union.